src/hitchhiker/tree/tracing_gc.cljc

```diff
@@ -4,45 +4,40 @@
             [hitchhiker.tree.node :as n]
             [hitchhiker.tree.utils.async :as ha]))
 
-(defprotocol IGCScratch
-  (observe-addr! [this addr] "Marks the given addr as being currently active")
-  (observed? [this addr] "Returns true if the given addr was observed"))
+(defn mark
+  "Mark addresses in gc-roots, traversing the entire tree for each root.
+  If addresses is given, it must be a set containing possibly already-marked
+  addresses.
 
-(defmacro do-<!
-  "Force into an async taking call.
+  Returns all marked addresses."
+  ([gc-roots] (mark gc-roots #{}))
+  ([gc-roots addresses]
+   (loop [addresses (volatile! (transient addresses))
```

Comment:
Ok, one last thing, we do not need to use a mutable object for addresses and the loop here. I would just use a recursion and pure values. Actually we can even do it along the tree without the loop and merge the resulting sets of marks on the way up again, there is no risk of a stackoverflow. 

(Later we can collect go channels that return either marks or nothing and we can pipeline the reads in parallel that way.)

Reply:
Forget the pipelining comment, it is a bit trickier than that.

Reply:
I'm unsure what the remaining issue is with this implementation. If you would prefer it written some other way, feel free to change it yourself.

Reply:
Ok, sure.